### PR TITLE
Initial D3D9 port

### PR DESCRIPTION
**Pull request by LukeUsher**
This is the initial port of Cxbx-Reloaded to DirectX 9.

Notes:
1. The emulator still uses shader model 1.X at this point, migration to a newer shader model will be done in the next few weeks. (This is the main motivation for this port)
2. There will be very few improvements. Some Vertex Shaders that previously failed to compile now work, but the changes are minor at this point. We'll see major improvement once #1360 and #1361 are implemented.
3. There may be some regressions compared to develop: These will be ironed out as they are identified.

I'm going to merge this within the next few minutes: This isn't normal procedure, but we want this to be as well-tested as possible: The best way to do this is to make it easily available as an automated build.

Those who want stability should use the Release builds, rather than Dev builds, but those who want to test/experiment with the latest/greatest (like this feature!) can use Develop.

### CODE REVIEW

File: src/Common/Settings.cpp [hunk 1]
```diff
@@ -192,24 +191,32 @@ bool Settings::Init()
...
 		else {
 			return false;
 		}
 #endif
+		saveFile.append(szSettings_settings_file);
 
-		bRet = Save(saveFile);
+		// If the config file does not exists, create a blank one
+		// We can't call Save here because that overrides default values with false!
+		if (!std::experimental::filesystem::exists(saveFile)) {
```
Comment (RadWolfie) [inline src/Common/Settings.cpp]: This should not *have* experimental. There is a macro using experimental as std::filesystem.

File: src/Common/Settings.cpp [hunk 2]
```diff
@@ -58,9 +59,8 @@ Settings* g_Settings = nullptr;
 
 #define szSettings_setup_error "ERROR: Either setup have a problem or do not have write permission to directory."
 #define szSettings_init_error "ERROR: Unable to initialize Settings class."
-#define szSettings_save_user_option_message "If you want to save your settings in current/portable directory,\nclick 'Yes'." \
-                                            "\n\nIf you want to store your settings in user profile directory,\nclick 'No'." \
-                                            "\n\nClicking cancel will abort Cxbx-Reloaded."
+#define szSettings_save_user_option_message "Use Cxbx-Reloaded in Portable Mode?" \
```
Comment (RadWolfie) [inline src/Common/Settings.cpp]: The reason for depth explanation is so user from different country can understand it.

File: src/Common/Settings.cpp [hunk 3]
```diff
@@ -192,24 +191,32 @@ bool Settings::Init()
...
 			return false;
 		}
 #endif
+		saveFile.append(szSettings_settings_file);
 
-		bRet = Save(saveFile);
+		// If the config file does not exists, create a blank one
+		// We can't call Save here because that overrides default values with false!
+		if (!std::experimental::filesystem::exists(saveFile)) {
+			std::ofstream ofs(saveFile, std::ofstream::out);
```
Comment (RadWolfie) [inline src/Common/Settings.cpp]: Why ofstream?

File: src/Common/Settings.cpp (hunk 1 above)
Comment (LukeUsher) [reply src/Common/Settings.cpp]: hmm, well that macro didn't work here anyway: Once filesystem stops being experimental, we'll remove the experimental namespace.

File: src/Common/Settings.cpp (hunk 3 above)
Comment (LukeUsher) [reply src/Common/Settings.cpp]: fopen is a C function and unsafe, std::filesystem does not have "touch"/"create file" function, so we use an output filestream instead.

**Comment by mkultimate** [comment]
image

File: src/Common/Settings.cpp (hunk 3 above)
Comment (RadWolfie) [reply src/Common/Settings.cpp]: Also, I know why everything is set to blank/false. I removed the default values from a structure for not to re-set everything twice. One for default constructor, and second time for set the values.

**Comment by MarcoEstevez** [comment]
RalliSport Challenge - regresed

image

File: src/Common/Settings.cpp [hunk 4]
```diff
@@ -192,24 +191,32 @@ bool Settings::Init()
...
+			ofs << "\n";
+			ofs.close();			
+		}
 
-		// Check if saving a file is a success.
+		// Call LoadUserConfig, this will load the config, applying defaults for any missing fields
+		bRet = LoadUserConfig();
+
+		// Check if loading the file is a success.
+		bRet = true;
```
[the remark this answers is not in the document]
Comment (RadWolfie) [reply src/Common/Settings.cpp]: Ignore this, now it is checking if initialize default value is working correctly.

**Comment by RadWolfie** [comment]
Burnout title has a regression on boot.
```
Unhandled exception at 0x646C5853 (d3d9.dll) in Cxbx.exe:
0xC0000005: Access violation reading location 0x00000024.
```
Locate in CxbxKrnl/EmuD3D8.cpp, line 7306
https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/blob/99fb143a04aa03182311be54483498101a552d07/src/CxbxKrnl/EmuD3D8.cpp#L7306

I'm not sure what 0x24 is. Here's the argument values:
* ? (inline function)
* 0x00000002 = DrawContext.dwHostPrimitiveCount
* 0x07420530 = DrawContext.pHostVertexStreamZeroData
* 0x0000001C = DrawContext.uiHostVertexStreamZeroStride

**Pull request merged**